collector/mdadm.go

```diff
@@ -0,0 +1,200 @@
+// +build !nomdadm
+
+package collector
+
+import (
+	"io/ioutil"
+	"strconv"
+	"strings"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/log"
+)
+
+var statusfile = "/proc/mdstat"
+
+type mdstatus struct {
+	mdName       string
+	isActive     bool
+	disksActive  int64
+	disksTotal   int64
+	blocksTotal  int64
+	blocksSynced int64
+}
+
+type mdadmCollector struct {
+	isActive     *prometheus.GaugeVec
+	disksActive  *prometheus.GaugeVec
+	disksTotal   *prometheus.GaugeVec
+	blocksTotal  *prometheus.GaugeVec
+	blocksSynced *prometheus.GaugeVec
+}
+
+func init() {
+	Factories["mdadm"] = NewMdadmCollector
+}
+
+func evalStatusline(statusline string) (active, total, size int64) {
+
+	first_split := strings.Split(statusline, "[")
+	second_split := strings.Split(first_split[1], "]")
+
+	sizeDirty := strings.Split(first_split[0], "blocks")[0]
+	sizeStripped := strings.Replace(sizeDirty, " ", "", -1)
+
+	numberOfDrives := strings.Split(second_split[0], "/")
+
+	total, _ = strconv.ParseInt(numberOfDrives[0], 10, 64)
```

Comment:
Hmm, I agree that that would generally be ok normally. On the other hand, we'll want to move all procfs-parsing-related functionality into https://github.com/prometheus/procfs at some point, and the existing methods in that library like `NewLimits()`, `NewStat()` `NewIPVS()`, etc., just return an error immediately when something is wrong with the parsed file. So to keep it consistent, I'd still argue for an all-or-nothing approach. If we can't parse the entire file, in 99% of cases it would point to a bug in the parsing code here that should be fixed as soon as possible anyways. If we fail partially here, we'd also need a new metric to expose the partial failure to be able to alert on it. Total collector failures are currently already exposed via an existing metric here: https://github.com/prometheus/node_exporter/blob/master/node_exporter.go#L103